File: pytorch-torchscript/34_tuple.cpp
#include <torch/script.h>
#include <tuple>

static const std::string file = "../../34_tuple.torchscript";

void test(torch::Device device){
  torch::jit::script::Module module = torch::jit::load(file, device);

  std::cout << "named_attributes ===\n";
  for (const auto& attr : module.named_attributes()){
    std::cout << "attr: " << attr.name << "=" << attr.value << std::endl;
  }

  std::cout << "method.names ===\n";
  for (const auto& method : module.get_methods()){
    std::cout << method.name() << "\n";
  }

  at::Tensor a = at::ones({3, 1}, at::device(device).dtype(at::kInt));
  at::Tensor b = at::ones({3, 1}, at::device(device).dtype(at::kInt));

  std::vector<torch::jit::IValue> inputs;
  inputs.push_back(a);
  inputs.push_back(b);
  inputs.push_back(std::tuple<int, float, bool>{1, 1.1, true});

  auto output = module.forward(inputs);
  std::cout << output << std::endl;
}

int main() {
  auto cpu = torch::Device(torch::kCPU, 0);
  test(cpu);

  auto gpu = torch::Device(torch::kCUDA, 0);
  test(gpu);
}

File: pytorch-torchscript/34_tuple.py
# 근데 NamedTuple인 Opt를 C++에서 어떻게 만들지?
#from typing import NamedTuple
from typing import Tuple
import torch
import torch.jit

#Opt = namedtuple('Config', ['int_value', 'float_value', 'bool_value'])
# class MyOption(NamedTuple):
#     int_value: int
#     float_value: float
#     bool_value: bool


@torch.jit.script
def doit(a: torch.Tensor, b: torch.Tensor, option: Tuple[int, float, bool]):
    if option[2]:
        return a + b
    else:
        return a - b


if __name__ == '__main__':
    doit.save("34_tuple.torchscript")
    print(doit.code)

    #opt = MyOption(int_value=1, float_value=0.5, bool_value=False)
    result = doit(torch.tensor([1, 2, 3]), torch.tensor([4, 5, 6]), (1, 0.5, False))
    print(result)
